chisel/src/config.rs

```diff
@@ -0,0 +1,570 @@
+    // yaml: top-level yaml element generated by parser
+    fn from_yaml(yaml: &Value) -> Result<Self, String>
+    where
+        Self: Sized;
+}
+
+/// Main trait for generating a configuration usable by the driver from CLI arguments.
+pub trait FromArgs {
+    // Supported formatting example:
+    // modules: "remapimports, trimexports, verifyimports"
+    // options: "remapimports.preset=ewasm, trimexport.option=foo, verifyimports.a=b"
+    fn from_args(modules: &str, options: &str) -> Result<Self, String>
+    where
+        Self: Sized;
+}
+
+/// A list of rulesets for a single chisel run.
+/// Left-hand value: name of ruleset
+/// Right-hand value: ruleset configuration
+#[derive(Debug)]
+pub struct ChiselConfig(VecDeque<(String, Ruleset)>);
+
+/// A single ruleset, containing options and a set of module configurations.
+#[derive(Debug)]
+pub struct Ruleset {
+    /// Top-level ruleset configurations, such as the input/output file paths.
+    opts: HashMap<String, String>,
+    /// List of module configurations.
+    /// Left-hand value: name of module
+    /// Right-hand value: module options
+    modules: VecDeque<(String, ModuleConfig)>,
+}
+
+/// A key-value module configuration.
+#[derive(Debug)]
+pub struct ModuleConfig(HashMap<String, String>);
+
+impl ChiselConfig {
+    pub fn rulesets(&self) -> &VecDeque<(String, Ruleset)> {
+        &self.0
+    }
+
+    pub fn rulesets_mut(&mut self) -> &mut VecDeque<(String, Ruleset)> {
+        &mut self.0
+    }
+}
+
+impl Ruleset {
+    pub fn options(&self) -> &HashMap<String, String> {
+        &self.opts
+    }
+
+    pub fn options_mut(&mut self) -> &mut HashMap<String, String> {
+        &mut self.opts
+    }
+
+    pub fn modules(&self) -> &VecDeque<(String, ModuleConfig)> {
+        &self.modules
+    }
+
+    pub fn modules_mut(&mut self) -> &mut VecDeque<(String, ModuleConfig)> {
+        &mut self.modules
+    }
+}
+
+impl ModuleConfig {
+    pub fn options(&self) -> &HashMap<String, String> {
+        &self.0
+    }
+}
+
+impl FromArgs for ChiselConfig {
+    fn from_args(modules: &str, options: &str) -> Result<Self, String>
+    where
+        Self: Sized,
+    {
+        let mut ret = VecDeque::new();
+        match Ruleset::from_args(modules, options) {
+            Ok(rs) => {
+                ret.push_back(("cli".to_string(), rs));
+                Ok(ChiselConfig(ret))
+            }
+            Err(e) => Err(e),
+        }
+    }
+}
+
+impl FromArgs for Ruleset {
+    fn from_args(modules: &str, options: &str) -> Result<Self, String>
+    where
+        Self: Sized,
+    {
+        // Filter out whitespace for names.
+        let modules: String = modules.chars().filter(|c| *c != ' ').collect();
+
+        // Collect module names.
+        let module_names: Vec<String> = modules
+            .split(',')
+            .filter_map(|c| match c {
+                "" => None, // If the user put a comma somewhere weird, we need to filter out the empty string.
```

Comment:
Shouldn't this be an error instead?

Reply:
it's in a filter map expression, so it should be `Some` or `None`. This shouldn't be an error case tho.